lib/onelogin/ruby-saml/utils.rb

```diff
@@ -15,7 +15,21 @@ class Utils
 
       DSIG      = "http://www.w3.org/2000/09/xmldsig#"
       XENC      = "http://www.w3.org/2001/04/xmlenc#"
-      DURATION_FORMAT = %r(^(-?)P(?:(?:(?:(\d+)Y)?(?:(\d+)M)?(?:(\d+)D)?(?:T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?)?)|(?:(\d+)W))$)
+      DURATION_FORMAT = %r(^
+        (-?)P                         # 1: Duration sign
+        (?:
+          (?:(\d+)Y)?                 # 2: Years
```

Comment:
According to [Wikipedia](https://en.wikipedia.org/wiki/ISO_8601) all values (years, months, etc.) may have fractions.

Reply:
In [S4.4.3.2](https://www.loc.gov/standards/datetime/iso-tc154-wg5_n0038_iso_wd_8601-1_2016-02-16.pdf):

> If necessary for a particular application, the lowest order components may have a decimal fraction.

Which is vague, but I think means that only the H/M/S components may be fractional?

Reply:
"Lowest order" means the smallest value in the string. So `P0.5Y` or `P1Y0.5M` are both valid.

However, following the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle), and for code cleanliness reasons, let's just support decimal on all values, since there is no harm/ambiguity in doing so.

Reply:
I wondered about that interpretation, but the plural in "lowest order components" makes me doubt it. I also noted in the commit message that the `DateTime#next_year` / `#next_month` and `#next_day` methods that we're using to compute the duration for these components ignore any fractional parts, so even if we parse these components as decimals, we can't compute their value without a lot more changes.

I'm also not certain that fractional increments of a nominal duration (Y/M/W/D) are easily interpretable. For example, `DateTime#next_month` computes the same day of the month, N calendar months in the future. This has a clearly defined meaning, and there are still complications due to different-length months. E.g. Jan 31 + 1 month = Feb 28 (or Feb 29 on a leap year). What should the result be if we added half a month instead? Damned if I know. 😄

So unless we genuinely support fractional years/months/days, we're not following the robustness principle, and there is a harm in parsing them: clients sending e.g. `P0.5Y` will get a duration of zero instead of an informative error.

Reply:
Yeah seems reasonable to support fractional Hour, Minute, and Day as the others are non-deterministic. Even Day and Week can be different than 24h on Daylight Saving Time transitions, so there is potential ambiguity which is better to reject.

That being said, we should allow each one of H/M/S to have a fraction `PT0.5H0.5M0.1S`

Reply:
Thanks - yep, I think the code as written will pick up the fractions on any or all of those components.